### ISSUE TITLE
How to include .so in windows application

### ISSUE THREAD

**Issue by ReniDelonzek** [flutter/flutter#59515]
I want execute a native library in windows across DynamicLibrary.open(), but I not found  recommendations about make this. 

I am tried include .so in assets files and they are included in profile/release compilation, but the path in debug/release are distinct

Debug 
`build/flutter_assets/assets/file.so`

Release 
`data/flutter_assets/assets/file.so`

A possibility is change path dynamically, but I believe that there is a more suitable solution

**Comment by stuartmorgan** [comment]
Those look like relative paths from different locations. Can you give the exact paths from your project root that you are comparing?

**Comment by ReniDelonzek** [comment]
The root path is obtained through Directory.current.path, and then I concatenate with the rest.

I believe that there should be some way to recover the files (or even the correct path) present in the assets through AssetManager, however I verified that this is not possible today, I am looking for some other suitable solution

**Comment by jonahwilliams** [comment]
You can use the asset bundle to load your AssetManifest, have you tried that?

**Comment by ReniDelonzek** [comment]
@jonahwilliams  Can you share documentation on that or even an example for me to try?

**Comment by jonahwilliams** [comment]
Actually, you might not need that. Assuming you know the relative path to the library (which is what you provided in the asset manifest). Do you need the absolute path? If so you could run something like:

var absPath = File(assetPath).absolute.path;

**Comment by ReniDelonzek** [comment]
This solution can work in debug mode (considering that I know the file's location will be in assets/ from the execution directory). But the Windows Profile/Release build creates a different structure than that, where the asset path follows as I mentioned in the first example `data/flutter_assets/assets/` 
So the problem comes down to the fact that the relative path of assets is different in different releases

**Comment by jonahwilliams** [comment]
the file paths will be different on different machines too, that is why the asset paths are relative to `flutter_assets`. It seems like we're missing an API to have the asset bundle potentially return either the embedder's flutter_assets path or the full path to an asset

**Comment by ReniDelonzek** [comment]
Yes, my opinion is that there should be an API to return the path (relative or complete) of the assets, regardless of the platform/release that is being executed, that solves precisely this problem that I am trying to solve

**Comment by stuartmorgan** [comment]
> But the Windows Profile/Release build creates a different structure than that, where the asset path follows as I mentioned in the first example `data/flutter_assets/assets/`
> So the problem comes down to the fact that the relative path of assets is different in different releases

This is the part I'm confused by. You didn't answer my question above: what are the relative paths you've quoted there relative *to*? Because it looks like one is relative to the flutter project root, and the other is relative to the build output, unless I'm missing something.

> The root path is obtained through Directory.current.path

To clarify, I'm not asking about runtime, I'm asking how you got the two relative paths you listed in the original report, and what they are relative to on your filesystem.

Maybe a clearer way to ask this would be: can you please give the absolute paths on your machine (replacing any personal information with placeholders) for the debug and release asset paths?

**Comment by ReniDelonzek** [comment]
Of course, I will try to explain in a more detailed way, forgive my English.

As long as I include the file in the assets folder, its absolute path will be
`C:\Users\{user}\Desktop\timber_track\assets\file.so`

Then I would be able to get the full path to passing `DynamicLibrary.open` through `Directory.current.path + "assets\file.so"`

When I compile (`flutter build windows`), the resulting structure from the executable directory is:
`data\flutter_assets\assets\file.so`

So to get the path to the executable I would have to call `Directory.current.path + "data\flutter_assets\assets\file.so"`

As an alternative to the first situation (Debug Mode), the build tool generates in the \build folder a structure exactly identical to which is found in the Release version (flutter_assets\assets\file.so), therefore, the only difference between the Debug/Release versions is that in the debug I need to call `build\flutter_assets\...` and in the release `data\flutter_assets\...`

With all this, I will summarize my questions:

1. Is it correct to include an executable inside the assets folder, since I will only use it in the Windows application? Will this file be excluded from the builds for other platforms?

2. If the first is correct, what is the most appropriate way to recover the path of this file?

**Comment by stuartmorgan** [comment]
> its absolute path will be `C:\Users\{user}\Desktop\timber_track\assets\file.so`

This is its source location. You should never be using a source location in your build.

> Then I would be able to get the full path to passing `DynamicLibrary.open` through `Directory.current.path + "assets\file.so"`

You should never assume that the value of `Directory.current` is related to the location of your executable.

> When I compile (`flutter build windows`), the resulting structure from the executable directory is:
> `data\flutter_assets\assets\file.so`

You've switched back to a relative path again. I am assuming what you mean here is that it's at `C:\Users\{user}\Desktop\timber_track\build\windows\x64\Release\Runner\data\flutter_assets\...`?

> So to get the path to the executable I would have to call `Directory.current.path + "data\flutter_assets\assets\file.so"`

Per the above, you shouldn't do this either. `Directory.current` is not the same thing as the executable's location.

> As an alternative to the first situation (Debug Mode), the build tool generates in the \build folder a structure exactly identical to which is found in the Release version (flutter_assets\assets\file.so)

This is where I keep getting lost. What you're calling "debug mode" is the source location. The debug build's output is `C:\Users\{user}\Desktop\timber_track\build\windows\x64\Release\Runner\data\flutter_assets\...`

Are you saying that folder doesn't have your asset?

> Will this file be excluded from the builds for other platforms?

No, assets are bundled on all platforms.

**Comment by ReniDelonzek** [comment]
According to your answers, I believe that my way of resolving this is incorrect and should be discarded.

I believe that my question will one day be addressed here https://flutter.dev/docs/development/platform-integration/c-interop, but I understand that it is not ready yet

Anyway, my question still persists in understanding what is the most appropriate place to include the .so file, so that it is included only for the Windows compilation. And how should I retrieve his path to DynamicLibrary.open

**Comment by stuartmorgan** [comment]
Filed #61962 for the technical work needed to support this; leaving this open for documentation once that's done.
